### PR TITLE
Neural Network widget

### PR DESCRIPTION
**Pull request by ajdapretnar**
##### Issue

Neural Networks only in scripting part, not in a widget.

##### Description of changes
Added a simple, cumbersome NN widget. Should work for basic tasks. Documentation pending until layout consolidated.

##### Includes
- [X] Code changes
- [X] Tests
- [x] Documentation

### CODE REVIEW

File: Orange/widgets/model/owneuralnetwork.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from Orange.data import Table
+from Orange.modelling import NNLearner
+from Orange.widgets import gui
+from Orange.widgets.settings import Setting
+from Orange.widgets.utils.owlearnerwidget import OWBaseLearner
+
+
+class OWNNLearner(OWBaseLearner):
+    name = "Neural Network"
+    description = "Predict according to the nearest training instances."
```
Comment (kernc) [inline Orange/widgets/model/owneuralnetwork.py]: Update description.

File: Orange/widgets/model/owneuralnetwork.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    def add_main_layout(self):
+        box = gui.vBox(self.controlArea, "Network")
+        self.hidden_layers_edit = gui.lineEdit(box, self,
+                                               "hidden_layers_input",
+                                               label="Hidden layers:",
+                                               orientation=Qt.Horizontal,
+                                               callback=self.settings_changed)
+        self.activation_combo = gui.comboBox(
+            box, self, "activation_index", orientation=Qt.Horizontal,
+            label="Activation:", items=[i.capitalize() for i in self.activation],
```
Comment (kernc) [inline Orange/widgets/model/owneuralnetwork.py]: The correct human-friendly labels would be: Identity, Logistic, tanh and ReLU. Imho `.upper()` would server better.

File: Orange/widgets/model/owneuralnetwork.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            callback=self.settings_changed)
+        self.solver_combo = gui.comboBox(
+            box, self, "solver_index", orientation=Qt.Horizontal,
+            label="Solver:", items=[i.capitalize() for i in self.solver],
```
Comment (kernc) [inline Orange/widgets/model/owneuralnetwork.py]: `.upper()` would server better. Those are acronyms.

File: Orange/widgets/model/owneuralnetwork.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    alpha = Setting(0.0001)
+    max_iterations = Setting(200)
+
+    def add_main_layout(self):
+        box = gui.vBox(self.controlArea, "Network")
+        self.hidden_layers_edit = gui.lineEdit(box, self,
+                                               "hidden_layers_input",
+                                               label="Hidden layers:",
+                                               orientation=Qt.Horizontal,
+                                               callback=self.settings_changed)
```
Comment (kernc) [inline Orange/widgets/model/owneuralnetwork.py]: Add `tooltip=` describing its format.

File: Orange/widgets/model/owneuralnetwork.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    activation_index = Setting(3)
+    solver_index = Setting(2)
+    alpha = Setting(0.0001)
+    max_iterations = Setting(200)
+
+    def add_main_layout(self):
+        box = gui.vBox(self.controlArea, "Network")
+        self.hidden_layers_edit = gui.lineEdit(box, self,
+                                               "hidden_layers_input",
+                                               label="Hidden layers:",
```
Comment (kernc) [inline Orange/widgets/model/owneuralnetwork.py]: Neurons per hidden layer?

File: Orange/widgets/model/owneuralnetwork.py (hunk 1 above)
Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Ups.

File: Orange/widgets/model/owneuralnetwork.py (hunk 5 above)
Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Yeah, I'll update the label.

**Comment by ajdapretnar** [comment]
@kernc Would you care to check whether spin box maxval and steps make sense or not? To adjust this as well, while I'm at it.

**Comment by kernc** [comment]
I'd never wait to scroll through [1e-5, 1] in 1e-5 increments. Perhaps 1e-2 stepsize makes more sense in general, but the majority of users will just edit the number manually.

**Comment by ajdapretnar** [comment]
All updated.

File: Orange/widgets/model/owneuralnetwork.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        box = gui.vBox(self.controlArea, "Network")
+        self.hidden_layers_edit = gui.lineEdit(box, self,
+                                               "hidden_layers_input",
+                                               label="Neurons per hidden "
+                                                     "layer:",
+                                               orientation=Qt.Horizontal,
+                                               callback=self.settings_changed,
+                                               tooltip="A list of integers "
+                                                       "defining neurons. "
+                                                       "Lenght of list "
```
Comment (kernc) [inline Orange/widgets/model/owneuralnetwork.py]: typo. Also might be better to indent these params a single level, like below.

File: Orange/widgets/model/icons/NN.svg [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	l-0.819,1.352l-5.911-5.792c0.063-0.185,0.128-0.37,0.164-0.568c0.167-0.916-0.007-1.808-0.41-2.572l4.481-0.017l-0.308,1.201
+	l3.961-1.323l-2.834-3.07l-0.562,2.191l-5.458,0.021c-0.552-0.544-1.266-0.934-2.084-1.084c-2.172-0.398-4.256,1.039-4.653,3.211
+	c-0.399,2.176,1.041,4.261,3.212,4.656c0.969,0.178,1.913-0.025,2.705-0.481l4.805,7.734l-6.347,8.857
+	c-2.068-0.231-3.995,1.155-4.374,3.233c-0.399,2.176,1.041,4.261,3.212,4.656c1.112,0.204,2.197-0.078,3.048-0.688l5.882-0.022
+	l0.084,1.77l3.396-2.431l-3.612-2.101L32.628,38.258z M27.257,14.891c0.192-0.188,0.368-0.393,0.521-0.619l5.812,5.694l-1.703,2.376
+	L27.257,14.891z M33.717,27.185l-1.173,0.067l1.002,1.371l-6.161,4.698c-0.393-0.423-0.869-0.772-1.424-0.996l5.865-8.184
+	L33.717,27.185z"/>
+<path fill="#989898" d="M28.403,24.722l4.067-0.016l0.084,1.769l3.396-2.431l-3.612-2.101l0.084,1.762l-3.985,0.016
+	C28.464,24.05,28.462,24.383,28.403,24.722z"/>
+</svg>
```
Comment (janezd) [inline Orange/widgets/model/icons/NN.svg]: Icon is kind of crowded, lines are too thin... Maybe remove some arrows, and reduce the right-most layer to two neurons.

File: Orange/widgets/model/owneuralnetwork.py (hunk 6 above)
Comment (janezd) [reply Orange/widgets/model/owneuralnetwork.py]: +1

File: Orange/widgets/model/owneuralnetwork.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                                                       "defines number of "
+                                                       "layers. E.g. 4, 2, 2, 3.")
+        self.activation_combo = gui.comboBox(
+            box, self, "activation_index", orientation=Qt.Horizontal,
+            label="Activation:", items=[i.upper() for i in self.activation],
```
Comment (janezd) [inline Orange/widgets/model/owneuralnetwork.py]: Don't use `upper`. It looks ugly. If you'd like some -- but not all -- letters in upper case, add a list of labels in parallel with the method names you have now.

File: Orange/widgets/model/owneuralnetwork.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            callback=self.settings_changed)
+        self.solver_combo = gui.comboBox(
+            box, self, "solver_index", orientation=Qt.Horizontal,
+            label="Solver:", items=[i.upper() for i in self.solver],
+            callback=self.settings_changed)
+        self.alpha_spin = gui.doubleSpin(
+            box, self, "alpha", 1e-5, 1.0, 1e-2,
+            label="Alpha:", decimals=5, alignment=Qt.AlignRight,
+            callback=self.settings_changed, controlWidth=80)
+        self.max_iter_spin = gui.spin(
```
Comment (janezd) [inline Orange/widgets/model/owneuralnetwork.py]: Right-align.

File: Orange/widgets/model/owneuralnetwork.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            activation=self.activation[self.activation_index],
+            solver=self.solver[self.solver_index],
+            alpha=self.alpha,
+            max_iter=self.max_iterations,
+            preprocessors=self.preprocessors)
+
+    def get_learner_parameters(self):
+        return (("Hidden layers", self.get_hidden_layers()),
+                ("Activation", self.activation[self.activation_index].capitalize()),
+                ("Solver", self.solver[self.solver_index].capitalize()),
```
Comment (janezd) [inline Orange/widgets/model/owneuralnetwork.py]: `capitalize` or `upper` or a list of user-readable names.

File: Orange/widgets/model/owneuralnetwork.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        return self.LEARNER(
+            hidden_layer_sizes=self.get_hidden_layers(),
+            activation=self.activation[self.activation_index],
+            solver=self.solver[self.solver_index],
+            alpha=self.alpha,
+            max_iter=self.max_iterations,
+            preprocessors=self.preprocessors)
+
+    def get_learner_parameters(self):
+        return (("Hidden layers", self.get_hidden_layers()),
```
Comment (janezd) [inline Orange/widgets/model/owneuralnetwork.py]: This prints out as: **Hidden layers:** (100,)

File: Orange/tests/test_neuralnetwork.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+# pylint: disable=missing-docstring
+
+import unittest
+
+from Orange.data import Table
+from Orange.modelling import NNLearner, ConstantLearner
+from Orange.evaluation import CA, CrossValidation, MSE
+
+
+class TestKNNLearner(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (ajdapretnar) [reply Orange/tests/test_neuralnetwork.py]: Busted!

File: Orange/widgets/model/icons/NN.svg (hunk 7 above)
Comment (ajdapretnar) [reply Orange/widgets/model/icons/NN.svg]: Icon is taken from Orange2 and will be replaced with something nicer from Lara. Soon. :)

File: Orange/widgets/model/owneuralnetwork.py (hunk 8 above)
Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Hm, ok.

File: Orange/widgets/model/owneuralnetwork.py (hunk 11 above)
Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Yes. It should. I think.

File: Orange/widgets/model/owneuralnetwork.py (hunk 9 above)
Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Do you mean Qt.AlignRight as parameter?

File: Orange/widgets/model/owneuralnetwork.py (hunk 11 above)
Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Oh, yes, I see what's the issue.

Comment (ajdapretnar) [reply Orange/widgets/model/owneuralnetwork.py]: Oh no, I don't.

File: Orange/widgets/model/owneuralnetwork.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    solver_index = Setting(2)
+    alpha = Setting(0.0001)
+    max_iterations = Setting(200)
+
+    def add_main_layout(self):
+        box = gui.vBox(self.controlArea, "Network")
+        self.hidden_layers_edit = gui.lineEdit(box, self,
+                                               "hidden_layers_input",
+                                               label="Neurons per hidden "
+                                                     "layer:",
```
Comment (janezd) [inline Orange/widgets/model/owneuralnetwork.py]: Perhaps layer -> layers? It is not obvious the user can input a list. Not many users will find the tooltip.

File: Orange/widgets/model/owneuralnetwork.py (hunk 11 above)
Comment (janezd) [reply Orange/widgets/model/owneuralnetwork.py]: Something like `hidden_layer_sizes=", ".join(map(str, self.get_hidden_layers()))`

**Comment by ajdapretnar** [comment]
I don't think the widget should have an empty string as a default, especially since one can't use it on the fly. All other widgets have sklearn's default values and so should this one.
Moreover, the user can see the default is "100". Not sure, perhaps I can set the default to "100,". Would this make more sense?

Everything else is fixed, once agreed I can add docs and call it a day. ;)

**Comment by kernc** [comment]
An empty string results in no hidden layers. Such a perceptron is not forbidden by sklearn.

Set the default to "100,", but _also_ set the placeholder?

File: Orange/widgets/model/owneuralnetwork.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+    def get_learner_parameters(self):
+        return (("Hidden layers", ', '.join(map(str, self.get_hidden_layers()))),
+                ("Activation", self.act_lbl[self.activation_index]),
+                ("Solver", self.solv_lbl[self.solver_index]),
+                ("Alpha", self.alpha),
+                ("Max iterations", self.max_iterations))
+
+    def get_hidden_layers(self):
+        layers = tuple(map(int, re.findall('\d+', self.hidden_layers_input)))
```
Comment (astaric) [inline Orange/widgets/model/owneuralnetwork.py]: Lint suggests you use `r'\d+'`. I agree

**Comment by ajdapretnar** [comment]
Fixed and rebased.

**Comment by ajdapretnar** [comment]
I don't understand why tests don't pass. They pass locally.

**Comment by astaric** [comment]
It's not the tests that are failing. If you click on details, you can see that the pylint and documentation builds have failed.

[Pylint](https://travis-ci.org/biolab/orange3/jobs/274062985#L638) has two complaints:
  1. When you do MSE(), you get a scoring function, when the scoring function is given results, it returns a score. The way you should call it is MSE()(results) (or scorer = MSE(); scorer(results))
  2. It wants QApplication import at the top of the file.

[Documentation](https://travis-ci.org/biolab/orange3/jobs/274062986#L1780) complains about the images in the documentation not being as small as they could be.

**Comment by ajdapretnar** [comment]
I guess it's fixed now.

**Comment by ajdapretnar** [comment]
I'm eager to merge this. Please advise what to do with the rouge import line in \_\_main\_\_?
